src/PureScript/Backend/Chez/Convert.purs

```diff
@@ -118,8 +122,47 @@ codegenTopLevelBindingGroup
   :: CodegenEnv
   -> BackendBindingGroup Ident NeutralExpr
   -> Array ChezDefinition
-codegenTopLevelBindingGroup codegenEnv { bindings } =
-  Array.concatMap (codegenTopLevelBinding codegenEnv) bindings
+codegenTopLevelBindingGroup codegenEnv { bindings, recursive }
+  | recursive, Just bindings' <- NonEmptyArray.fromArray bindings = do
+      -- Turn non-lazy bindings into lazy bindings so that they can refer to each
+      -- other during initialization. Some definitions are initialized lazily,
+      -- like lambda exressions, so we don't have to touch those.
+      let lazyBindings = NonEmptyArray.partition isLazyBinding bindings'
+      let
+        codegenEnv' = codegenEnv
+          { lazyTopLevelRefs = foldMap (Set.singleton <<< fst) lazyBindings.no }
+      fold
+        [ Array.concatMap (codegenTopLevelBinding codegenEnv') lazyBindings.yes
+        , map (codegenLazyTopLevelBinding codegenEnv') lazyBindings.no
+        , map (\(Tuple (Ident ident) _) -> Define ident $ S.forceLazyRef ident) lazyBindings.no
+        ]
+  | otherwise = Array.concatMap (codegenTopLevelBinding codegenEnv) bindings
+
+isLazyBinding :: Tuple Ident NeutralExpr -> Boolean
+isLazyBinding (Tuple _ expr) = case unwrap expr of
```

Comment:
I wonder if I should be more aggressive here @natefaubion?

Reply:
By aggressive, you mean try harder to detect more lazy bindings? It's not really an issue of correctness (you can make every binding in every recursive group lazy). You are are just potentially eliminating overhead where it's not needed. The one for the ES backend does work a little harder, yes, but not much.
https://github.com/aristanetworks/purescript-backend-optimizer/blob/bdef8b58cd8591bef7fd499c892989604e6b918d/backend-es/src/PureScript/Backend/Optimizer/Codegen/EcmaScript/Convert.purs#L802-L854